saved_snippets: Add keyboard UI for inserting a saved snippet.

Vector73 opened the pull request:
This is a follow-up of https://github.com/zulip/zulip/pull/31359.
- The first commit moves "Saved snippets" button from "send_later" popover to message formatting buttons and adds support to use saved snippets in message-edit UI.
- The second commit adds `Ctrl+'` keyboard shortcut to open saved snippet dropdown when the target textarea is in focus.

Fixes #31829.
Fixes #31830.

<!-- Describe your pull request here.-->



<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details><summary>**Screenshots and screen captures:**</summary>
<p>

![image](https://github.com/user-attachments/assets/e506b07a-54b3-4548-b3a3-e262027f00e3)

![image](https://github.com/user-attachments/assets/c4a779f2-e9e2-4115-bbe8-6030bad638e5)


</p>
</details> 


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
Do we want the popover to appear next to the cursor, or centered? It feels weird having it appear next to the send button.

alya commented:
Centered (and without the menu underneath) is what I suggested on CZO.

timabbott commented:
I think the next step for this is to rebase on top of https://github.com/zulip/zulip/pull/32988, which now lets us add additional widgets without undue pain.

alya commented:
Nice! A couple of comments from playing with it:

1. The tooltip should include the keyboard shortcut, as usual. Like this:
![Screenshot 2025-01-16 at 11 56 55@2x](https://github.com/user-attachments/assets/a4ea65c0-e55d-4aa6-94dd-be65a24d8218)
1. The icon looks much too big compared to the rest, and I think the stroke might be too thick as well. You can check with @terpimost if you aren't sure how to fix it.
![Screenshot 2025-01-16 at 11 58 45@2x](https://github.com/user-attachments/assets/f3f95550-1a0a-40c6-be32-f3ed03a3d229)

terpimost commented:
I think icons in compose box should be 22x22. Also I think empty message isn't a great icon for a snippet, which signalize that message would be filled with something. Here is 22x22 icon out of modified lucid icon:
<img width="318" alt="image" src="https://github.com/user-attachments/assets/6f8a0321-b9d7-4a08-b62d-c4560dc56759" />

```
<svg xmlns="http://www.w3.org/2000/svg" width="22" height="22" fill="none" viewBox="0 0 22 22">
  <g fill="#666">
    <path d="M6.877 8.217a.68.68 0 0 1 .681-.681h4.086a.68.68 0 1 1 0 1.361H7.558a.681.681 0 0 1-.68-.68Zm.681 2.043h6.81a.681.681 0 0 1 0 1.361h-6.81a.681.681 0 0 1 0-1.362Z"/>
    <path d="M3.789 4.288a2.043 2.043 0 0 1 1.444-.598h11.534a2.043 2.043 0 0 1 2.043 2.043v7.81a2.043 2.043 0 0 1-2.043 2.042h-9.89L4.353 18.11a.681.681 0 0 1-1.163-.481V5.733c0-.542.216-1.062.599-1.445Zm1.444.764a.681.681 0 0 0-.68.68v10.252l1.56-1.561a.68.68 0 0 1 .482-.2h10.172a.681.681 0 0 0 .68-.68v-7.81a.681.681 0 0 0-.68-.681H5.233Z"/>
  </g>
</svg>
```

Vector73 commented:
Added keyboard shortcut in the tooltip and fixed the icon:
![image](https://github.com/user-attachments/assets/478e8a27-a0ab-4836-ac9f-6d86f99b3e66)

alya commented:
Cool, @karlstolley are you up for reviewing this next?

timabbott commented:
A few small notes from testing this:
- When you use `Ctrl+'` to open the menu and select "Add a new saved snippet" with the keyboard, keyboard focus doesn't seem to be initially placed in the input areas; you have to hit `Tab` to get there. 
- If you type "t" in the filter widget, that appears to still match "add a new saved snippet", but then disappears after I type "ta", which makes me think we're string-matching on the actual label string. I think probably that special string shouldn't match after you start typing something, given it's at the top for an empty input? The other sane option would be for it to match any input, but be sorted below any existing saved inputs, so you can type a saved input name idea and hit enter and land in the modal with the name prepopulated. My temptation is to have it not match anything typed; it seems less complex and satisfactory.
- The new keyboard shortcut is missing documentation; it should go in the "composing messages" section.

I'm going to test-deploy on chat.zulip.org for UX feedback now.

alya commented:
We could follow GitHub and put the new snippet option at the bottom in a sticky area.

I think bottom placement is pretty good (as long as it's visible), in that hopefully most of the time you're using an existing snippet, not making a new one.

alya commented:
> The new keyboard shortcut is missing documentation; it should go in the "composing messages" section.

I guess after "Insert link" is fine.

Vector73 commented:
> We could follow GitHub and put the new snippet option at the bottom in a sticky area.

So something like this? I am not sure about the color though. 
![image](https://github.com/user-attachments/assets/ec3adb2d-3143-43f0-8464-528ae496e26f)

alya commented:
Yeah, that's what I was thinking. I guess on GitHub, the Tab keyboard navigation is filter -> scrollable options (if any) -> sticky option, which seems good.

alya commented:
For the color, I think we can make it look similar to sticky areas in scrollable modals. This is from the move messages modal when the screen is short.
![Screenshot 2025-01-23 at 10 13 37](https://github.com/user-attachments/assets/d8e882e3-fd88-4d96-af5f-2ae7fc5e8e5d)
![Screenshot 2025-01-23 at 10 12 55](https://github.com/user-attachments/assets/c1bf6275-c9f8-455d-816a-9e1035a2fa94)

alya commented:
(Though we do expect to do a design pass on those soon.)

alya commented:
And let's make the "Add" -> "Create" label tweak on that menu option.

karlstolley commented on web/templates/compose_control_buttons.hbs:
```diff
@@ -32,6 +32,9 @@
         <a role="button" class="compose_control_button compose_gif_icon zulip-icon zulip-icon-gif" aria-label="{{t 'Add GIF' }}" tabindex=0></a>
     </div>
+    <div class="compose_control_button_container preview_mode_disabled" data-tooltip-template-id="add-saved-snippet-tooltip">
+        <a id="saved_snippets_widget" role="button" class="compose_control_button zulip-icon zulip-icon-message-square-text" aria-label="{{t 'Add saved snippet' }}" tabindex=0></a>
```
Can this be refactored without the `saved_snippets_widget` ID? Because this can appear both in the compose box and the message-edit form, it would be better to use a unique class here.

Vector73 replied:
The dropdown widget binds to the saved snippet button using the ID of the button:
```js
constructor(options: DropdownWidgetOptions) {
    this.widget_name = options.widget_name;
    this.widget_selector = `#${CSS.escape(this.widget_name)}_widget`;
    // A widget wrapper may not exist based on the UI requirement.
    this.widget_wrapper_id = `${this.widget_selector}_wrapper`;
    this.widget_value_selector = `${this.widget_selector} .dropdown_widget_value`;
```

If we want to change this so that it binds to the button using the class name, we will need to refactor all the dropdown widgets. Can we write specialized logic to handle just the `saved_snippets_widget` in `dropdown_widget.ts`?

Vector73 replied:
I have added custom logic to use class name instead of ID to bind dropdown to `saved_snippets_widget`:
```js
        this.widget_name = options.widget_name;
        this.widget_selector = `#${CSS.escape(this.widget_name)}_widget`;
        if (this.widget_name === "saved_snippets") {
            // Since multiple `saved_snippets_widget` buttons can exist, we want them
            // to be referenced by the class name rather than id to avoid multiple
            // elements with the same id.
            this.widget_selector = `.${CSS.escape(this.widget_name)}_widget`;
        }
 ```

karlstolley replied:
That's not the prettiest thing I've ever seen, but I'd think an improvement would be to add `widget_selector` to the options, and then get out of the business entirely of trying to infer that from `widget_name`.

karlstolley replied:
(That's probably outside the scope of this PR though)

Vector73 replied:
Yeah, it would be much better to provide `widget_selector` as an optional argument to the constructor and it would be passed only for the saved snippets dropdown (`".saved_snippets_widget"`). In other dropdowns where this argument is not passed, we can fallback to the original logic of inferring `widget_selector` from `widget_name`.

Vector73 replied:
Updated the PR with the above change.

karlstolley commented on web/src/hotkey.js:
```diff
@@ -878,6 +879,13 @@ export function process_hotkey(e, hotkey) {
+        if (event_name === "open_saved_snippet_dropdown") {
+            const $messagebox = $(":focus").parents(".messagebox");
+            if ($messagebox.length === 1) {
+                $messagebox.find("#saved_snippets_widget")[0].click();
```
A class would indeed read better here, as IDs are supposed to be wholly unique.

Vector73 commented:
Updates made in the recent changes:
- Renamed "Add a new saved snippet" to "Create a new saved snippet".
- Renamed "Add saved snippet" tooltip to "Insert saved snippet".
- Moved "Create a new saved snippet" to the bottom of the dropdown in a sticky area (see screenshots).
- Avoided multiple saved snippets button with same ID by binding the dropdown widget to the button using class name.

alya commented:
Thanks! Playing with the PR:

1. Keyboard nav doesn't work properly when there are no results for the current filter. Arrows don't work to get to "Create new snippet" in that situation, and Tab escapes the menu. Chatting with @timabbott , we can just have Tab do the same thing as down arrow in this menu (like it does when there are results).
2. The visuals for selecting the sticky option with the mouse or keyboard aren't quite working right. I think that should look the same as selecting any of the non-sticky options.

alya commented:
I do like having then new snippet option sticky at the bottom rather than scrolly at the top.

Vector73 commented:
> 1. Keyboard nav doesn't work properly when there are no results for the current filter. Arrows don't work to get to "Create new snippet" in that situation, and Tab escapes the menu. Chatting with @timabbott , we can just have Tab do the same thing as down arrow in this menu (like it does when there are results).
> 2. The visuals for selecting the sticky option with the mouse or keyboard aren't quite working right. I think that should look the same as selecting any of the non-sticky options.

- Fixed keyboard navigation for empty dropdown list.
- Fixed `hover` and `focus` styles for the sticky option to make it similar to other options in the dropdown list.

timabbott commented:
Looks great, merged, thanks @Vector73!
